src/event/metric.rs

```diff
@@ -347,49 +359,130 @@ impl Metric {
     pub fn delete_tag(&mut self, name: &str) -> Option<String> {
         self.series.tags.as_mut().and_then(|tags| tags.remove(name))
     }
+
+    /// Create a new metric from this with the data zeroed.
+    pub fn zero(&self) -> Self {
+        Self {
+            series: self.series.clone(),
+            data: self.data.zero(),
+        }
+    }
 }
 
 impl MetricData {
-    /// Create new MetricData from this with all the data but marked as absolute.
-    pub fn to_absolute(&self) -> Self {
+    /// Rewrite this data to mark it as absolute.
+    pub fn into_absolute(self) -> Self {
         Self {
             timestamp: self.timestamp,
             kind: MetricKind::Absolute,
-            value: self.value.clone(),
+            value: self.value,
+        }
+    }
+
+    /// Rewrite this data to mark it as incremental.
+    pub fn into_incremental(self) -> Self {
+        Self {
+            timestamp: self.timestamp,
+            kind: MetricKind::Incremental,
+            value: self.value,
         }
     }
 
     /// Update this MetricData by adding the value from another.
     pub fn update(&mut self, other: &Self) {
-        match (&mut self.value, &other.value) {
-            (MetricValue::Counter { ref mut value }, MetricValue::Counter { value: value2 }) => {
+        self.value.add(&other.value)
+    }
+
+    /// Add the data from the other metric to this one. The `other` must
+    /// be relative and contain the same value type as this one.
+    pub fn add(&mut self, other: &Self) {
```

Comment:
Would it make sense to return a `Result` type here? Currently it silently does nothing if the other type is not `::Incremental`, which could lead to bugs that are hard to track down.

Reply:
This calling interface is copied from the previous implementation. I opted not to change it. I agree it is less than ideal, and we should open an issue to improve this. The callers I was working with ensure the metric kind and type are valid before calling.

Reply:
Moved to https://github.com/timberio/vector/issues/6329